9161 py3imap4

glyph opened the pull request:
twisted.mail.imap4 now works on Python 3.

glyph committed:
a few tweaks to get started

glyph committed:
Merge remote-tracking branch 'origin/trunk' into py3imap4

glyph committed:
helper functions & tests

glyph committed:
-2

glyph committed:
Merge remote-tracking branch 'origin/trunk' into py3imap4

glyph committed:
Merge remote-tracking branch 'origin/trunk' into py3imap4

glyph committed:
syntax errors on py27

glyph committed:
Preserve ordering for testing (how did this ever work before)

glyph committed:
news fragment

glyph committed:
remove bytesify placeholder that was only transitionally useful

rodrigc reviewed (commented):
I think you have a case in this patch of b"%s" % something which won't work on Python 3.3, which is still being used on the builders.

glyph commented:
> I think you have a case in this patch of b"%s" % something which won't work on Python 3.3, which is still being used on the builders.

Which builders?

glyph commented:
None of the build statuses on this PR are for 3.3, and I don't see a 3.3 on https://buildbot.twistedmatrix.com/boxes-supported?branch=trunk&num_builds=10 anywhere.

rodrigc commented:
.travis.yml on trunk has a python 3.3 build

glyph commented:
> .travis.yml on trunk has a python 3.3 build

Yes, but this is a configuration mistake that is irrelevant because we always explicitly specify which environment to run, no?  (As it happens I'm running the py33 tests locally and they pass.)

rodrigc commented:
The .travis.yml file explicitly specifies using python 3.3 for the 3.3 build.

glyph commented:
> The .travis.yml file explicitly specifies using python 3.3 for the 3.3 build.

Ah, I see the build now.  (It's still passing, FWIW, but that means the lines in question are uncovered. :))

rodrigc commented:
Lulz!

glyph committed:
ahem

glyph commented:
OK, tests are now failing (as they should be).

glyph commented:
This explains why the 3.3 build was passing despite invalid syntax ;)

markrwilliams committed:
Passes on Python 3: HandCraftedTests.testPathelogicalScatteringOfLiterals

markrwilliams committed:
Passes on Python 3: HandCraftedTests.test_authenticationChallengeDecodingException

markrwilliams committed:
Passes on Python 3: IMAP4ClientFetchTests.test_incompleteFetchSpecificTextResponse

markrwilliams committed:
Passes on Python 3: IMAP4ServerTests.testClose

markrwilliams committed:
Passes on Python 3: IMAP4ServerTests.testCreate

markrwilliams committed:
Passes on Python 3: IMAP4ServerTests.testDelete

MemoryAccount previously violated  RFC 3501 by allowing DELETEs to
remove mailboxes with hierarchically inferior names by comparing a
list to an integer.  This always returned True in Python 2, so
mailboxes were always deleted.

This commit also introduces and tests logic that prevents that
deletion.

markrwilliams committed:
Passes on Python 3: IMAP4ServerTests.testExpunge

markrwilliams committed:
Passes on Python 3: IMAP4ServerTests.testStatus

markrwilliams committed:
Passes on Python 3: test_imap.IMAP4ServerTests.testFullAppend

markrwilliams committed:
Passes on Python 3: test_imap.IMAP4ServerTests.testIllegalDelete

markrwilliams committed:
Passes on Python 3: test_imap.IMAP4ServerTests.testLSub, testList

MemoryAccount now stores mailbox names as text; _parseMbox was lifted
out of IMAP4Server so that MemoryAccount can use it.

The current API allows users to create and query mailboxes as text in
Python 3 via the imap4-utf-7 codec.  At the same time, the results are
returned as native strings on Python 2, so that a LIST response has
the mailbox and delimiter as bytes.  However, on Python 3,
IMAP4Client.list and .lsub should return text to maintain symmetry
with the outbound APIs.  This is now achieved with version-specific
decoding.

References should also be text, because they refer to mailboxes, which
are text.

markrwilliams committed:
Passes on Python 3: test_imap.IMAP4ServerTests.testLoginRequiringQuoting

Also log login errors to ease future debugging efforts.

markrwilliams committed:
Passes on Python 3: twisted.mail.test.test_imap.IMAP4ServerTests.testNamespace

The current API allows users to create and query mailboxes as text in
Python 3 via the imap4-utf-7 codec.  At the same time, results are
returned as native strings on Python 2, so that the namespace query
response has mailbox namespaces and delimiters as bytes.  However, on
Python 3, IMAP4Client.namespace return text to maintain symmetry with
the outbound APIs.  This is now achieved with version-specific
decoding.

markrwilliams committed:
Passes on Python 3: test_imap.IMAP4ServerTests.testNonExistentDelete

markrwilliams committed:
Passes on Python 3: test_imap.IMAP4ServerTests.testPartialAppend

markrwilliams committed:
Use inlineCallbacks for testFullAppend, too

markrwilliams committed:
Passes on Python 3: test_imap.IMAP4ServerTests.testRename

markrwilliams committed:
Passes on Python 3: test_imap.IMAP4ServerTests.testSelect

The keys in the returned dictionary are native strings because they
come from a fixed set of values.  In fact, they should eventually
become constantly constants.

markrwilliams committed:
Passes on Python 3: test_imap.IMAP4ServerTests.testStatus

IMAP4ServerTests round-trips with IMAP4Client, so IMAP4Client.status
has to encode STATUS names.  This commit limits STATUS names to those
defined in the RFC, which are all encodable as ASCII, to allow users
to continue to provide native strings while avoiding encoding issues.

In the event an IMAP server provides IMAP4Client with an undecodable
STATUS name, it will be logged and discarded.

Because STATUS names are now native strings,
IMAP4ServerTests.testFailedStatus has been been patched to pass native
string statuses.

markrwilliams committed:
Passes on Python 3: test_imap.IMAP4ServerTests.test_examine

markrwilliams committed:
Passes on Python 3: test_imap.testHierarchicalRename.

Mailbox names are native strings since
2fbedf0e97e6d75620b73927492ee623f8edf728.

markrwilliams committed:
Passes again on Python 3: testList.

Mailbox names are native strings since
2fbedf0e97e6d75620b73927492ee623f8edf728.  This broke comparisons
inside sortNest.  Eliminate this function in favor of a slower
order-independent equality check.

markrwilliams committed:
Passes again on Python 2: IMAP4ClientStatusTests

Sets and byte strings have different reprs in Python 2 and 3, but the
client tests introduced in cd63b86e0591793aa670f4eda50bb6fad0fac71d
assumed Python 3 reprs.

markrwilliams committed:
Remove PreauthIMAP4ClientMixin._extractDeferredResult.

All tests that inherit from this can use
SynchronousTestCase.successResultOf and
SynchronousTestCase.failureResultOf instead, with the added benefit
that the latter produces tracebacks that point at the original source
of the error.

markrwilliams committed:
Passes on Python 3: twisted.mail.test.test_imap.IMAP4ClientExaminTests,IMAP4ClientSelectTests

markrwilliams committed:
Passes on Python 3: twisted.mail.test.test_imap.IMAP4ClientExpungeTests

markrwilliams committed:
Passes on Python 3: twisted.mail.test.test_imap.IMAP4ClientSearchTests

markrwilliams committed:
Undecodable STATUS names should raise IllegalClientResponse.

Undecodable names are just another form of invalid responses, and
the status API already raises IllegalClientResponse in some cases.

Stop doing something different by logging and then discarding illegal
data!

markrwilliams committed:
Passes on Python 3: test_imap.HandCraftedTests

Returned protocol elements that are 7 bit safe are decoded as native
strings.  This means the protocol elements in the responses from
IMAP4Client.fetch methods should be native strings, while the returned
data should be bytes.

markrwilliams committed:
Passes on Python 3: test_imap.IMAP4ClientFetchTests

markrwilliams committed:
Passes on Python 3: test_imap.IMAP4ClientStoreTests

Code in this commit uses both "FLAGS" (native string) and b"FLAGS"
(byte string) to refer to a particular protocol element.

Native strings are used for protocol elements that are provided by a
user or returned to a user.

Byte strings are used when assembling a protocol message.

markrwilliams committed:
Passes on Python 3: twisted.mail.test.test_imap.NewStoreTests

markrwilliams committed:
Remove unused import.

markrwilliams committed:
Passes on Python 3: test_imap.NewFetchTests.testFetchFlags

markrwilliams committed:
Passes on Python 3: test_imap.NewFetchTests.testFetchFlagsUID

markrwilliams committed:
Passes on Python 3: test_imap.NewFetchTests.testFetchInternalDate

markrwilliams committed:
Passes on Python 3: test_imap.NewFetchTests.testFetchEnvelope

Note that IMAP4Server expects IMessage.getHeaders to return a
dictionary composed of native strings now.  There's a practical reason
and a conceptual reason for this.

The practical reason is that Python 3's email packages work in terms
of native strings, and most implementers of IMessage probably use
these.

The theoretical reason is that RFC 6532 allows UTF-8 in email headers,
and RFC 6855 extends IMAP4rev1 to support this.  imap4 can and should
support Unicode strings beyond those that are ASCII congruent Given an
RFC 6855 implementation.

IMAP4Client returns envelope data as native strings decoded with
charmap on Python 3.  This will cause mojibake, but maintains the
native-string API.

markrwilliams committed:
Merge branch 'trunk' into 9161-py3imap4

markrwilliams committed:
twisted.mail.imap_4.parsedIdList takes bytes.

It is essential to IMAP4Server, which must parse message sequences
bytes off the wire, and optional for IMAP4Client, whose public API can
use MessageSet.

markrwilliams committed:
Passes on Python 3: test_imap.NewFetchTests.testFetchAll

markrwilliams committed:
Passes on Python 3: test_imap.NewFetchTests.testFetchBody

markrwilliams committed:
Passes on Python 3: test_imap.NewFetchTests.testFetchBodyParts

markrwilliams committed:
Passes on Python 3: test_imap.NewFetchTests.testFetchSimplifiedBody

markrwilliams committed:
Passes on Python 3: test_imap.NewFetchTests.testFetchSimplifiedBodyRFC822

markrwilliams committed:
Passes on Python 3: test_imap.NewFetchTests.testFetchSimplifiedBodyText

markrwilliams committed:
Passes on Python 3: test_imap.NewFetchTests.NewFetchTests.test_fetchBodyStructure

markrwilliams committed:
Passes on Python 3: test_imap.NewFetchTests.NewFetchTests.testFetchFast

markrwilliams committed:
Passes on Python 3: test_imap.NewFetchTests.NewFetchTests.testFetchFull

markrwilliams committed:
Passes on Python 3: test_imap.NewFetchTests.testFetchMessage,
_formatHeaders expects native strings.

This is a consequence of IMessage.getHeaders return a dictionary of
native strings, introduced in
165dcd4997f758156163e15f20abdf420ff48b84.

markrwilliams committed:
Passes on Python 3: test_imap.NewFetchTests.NewFetchTests.testFetchHeaders

markrwilliams committed:
Passes on Python 3: test_imap.NewFetchTests.NewFetchTests.testFetchSize

markrwilliams committed:
Passes on Python 3: test_imap.NewFetchTests.test_fetchBodyPartOfNonMultipart

markrwilliams committed:
Passes on Python 3: test_imap.NewFetchTests.test_fetchBodyStructureMultipart

markrwilliams committed:
Passes on Python 3: test_imap.NewFetchTests.test_fetchSimplifiedBodyMultipart

markrwilliams committed:
Passes on Python 3: test_imap.IMAP4HelperTests.test_headerFormatter

A consequence of _formatHeaders taking a dictionary of native strings
and returning bytes, introduced in
451f2cf1d1a262fee5c82196c2ab0f1d0b5998f9

markrwilliams committed:
Passes on Python 3: test_imap.testUnfeaturefulMessage

A consequence of _formatHeaders taking a dictionary of native strings
and returning bytes, introduced in
451f2cf1d1a262fee5c82196c2ab0f1d0b5998f9

markrwilliams committed:
Passes on Python 3: test_imap.FetchSearchStoreTests.testSearch

markrwilliams committed:
Passes on Python 3: test_imap.MessageProducerTests

A consequence of _formatHeaders taking a dictionary of native strings
and returning bytes, introduced in
451f2cf1d1a262fee5c82196c2ab0f1d0b5998f9

markrwilliams committed:
Passes on Python 3: test_imap.TLSTests.testAPileOfThings

markrwilliams committed:
Merge branch 'trunk' into 9161-py3imap4

markrwilliams committed:
IMAP4Client fetch methods accept message sequences as native strings.

Message identifiers are ASCII, so they can be native strings.  This
eases porting and provides a more comfortable API for Python 3 users.

markrwilliams committed:
IMAP4Client.set, add, and removeFlags accept native-string message sequences.

The reasoning here is the same as in
d669fe4f5597e236ec3425d8874972745b9e682b.

markrwilliams committed:
IMAP4Client.copy accepts native-string message sequences.

The reasoning here is the same as in
d669fe4f5597e236ec3425d8874972745b9e682b.  This also introduces tests
to cover IMAP4Client.copy

markrwilliams committed:
IMAP4Client docstring clean up.

Use L instead of C; fetch et. al. accept strings or MessageSets, never
just bytes.

markrwilliams committed:
FETCH responses are native strings.

Section 7.4.2 of RFC 3501 allows 8-bit data only for BODY sections
with a charset identifier; everything else is 7 bits.  For Python 2,
all response components except BODY sections with charsets can be
native strings, whereas on Python 3, they can all be native strings
decoded as ASCII, *except* for BODY sections with CHARSETs.

Until https://twistedmatrix.com/trac/ticket/9201 is closed, Python 3
can side-step the issue by decoding fetch responses with charmap.
This will produce ASCII-congruent native strings in almost all cases,
and recoverable mojibake in the case a BODY's CHARSET was specified.

markrwilliams committed:
Don't reuse flags variable.

markrwilliams committed:
Flags are now consistently native strings.

markrwilliams committed:
IMAP4Client.search queries are native strings.

This is the existing API on Python 2.  Search terms are generally
7-bit except when a charset identifier identifies them as a separate
encoding.  Both to ease porting and to ensure the Python 3 API allows
arbitrarily encoded text when #9201 is closed and imap4 supports
charset identifiers, client search queries should be native strings
encoded with charmap.

markrwilliams committed:
IMAP4Client.fetchSpecific accepts headers as native strings.

The client should accept headers as native strings for the same
reasons as 165dcd4997f758156163e15f20abdf420ff48b84 has
IMessage.getHeaders return a dictionary of native strings.

markrwilliams committed:
IMAP4Client.search queries are only encoded on Python 3.

This ensures that arbitrary byte string queries continue to work on
Python 2.

markrwilliams committed:
Nuke unused variable.

markrwilliams committed:
Make imap4client example program work on Python 3.

rodrigc commented:
@markrwilliams can you look at: https://github.com/twisted/twisted/pull/738 ?  maybe we can combine efforts as you move forward with your porting work

markrwilliams committed:
IMAP4Server can parse empty string literals as final arguments.

This closes https://twistedmatrix.com/trac/ticket/9207 and improves
test coverage.

markrwilliams committed:
Remove unused MessageSet.__bytes__ method.

markrwilliams committed:
codecs.CodecInfo has existed since Python 2.6

markrwilliams committed:
Test unconvered paths in imap4-utf-7 decoder and encoder.

markrwilliams committed:
Test invalid inputs to imap4.parseTime.

markrwilliams committed:
100% coverage for imap4.FileProducer.

markrwilliams committed:
Coverage for imap4._FetchParser.

markrwilliams committed:
100% coverage for imap4.MessageProducer.

markrwilliams commented:
@rodrigc Sure!

markrwilliams committed:
Merge branch 'trunk' into 9161-py3imap4

markrwilliams committed:
Remove dead branches from imap4._getContentType.

The tests revealed that the provided branches would never run.  This
commit removes them in favor of less code with what should be
equivalent behavior.

markrwilliams committed:
100% coverage for imap4.Query and imap4.Or

markrwilliams committed:
IMAP4Client.fetchSpecific generates valid HEADER section.

Previously, the HEADER section (requested by
IMAP4Client.fetchSpecific(id, headerType="HEADER")) would contain an
empty parameter list.  This violates RFC 3501 (see
https://tools.ietf.org/html/rfc3501#section-6.4.5).

This also introduces 100% test coverage for IMAP4Client.fetchSpecific

markrwilliams committed:
Remove useless import.

markrwilliams committed:
Merge branch 'trunk' into 9161-py3imap4

markrwilliams committed:
Remove unused variable.

markrwilliams committed:
100% test coverage for LiteralString and LiteralFile.

This reverts the change to LiteralString introduced in
8f5ac1a0c2a543ea0d088c52bd068e49380d719a, because it incorrectly
handled the case where data's length is exactly the size of the
literal.

markrwilliams committed:
Sanity test for imap4.Command.__repr__

markrwilliams committed:
Tests for imap4.WriterBuffer.

markrwilliams committed:
Define IChallengeResponse, test custom challengers with
imap4.IMAP4Server.

A comment mentioned IChallengeResponse but it was never defined.
Define it, verify that twisted.mail's two credentials providers
provide this interface, and assert that IMAP4Server's constructor
accepts a dictionary of challengers.

markrwilliams committed:
Server timeouts close the selected mailbox.

markrwilliams committed:
100% coverage for IMAP4Server wire protocol parsing.

markrwilliams committed:
Merge branch 'trunk' into 9161-py3imap4

markrwilliams committed:
imap4: Test untagged response buffering.

Also, remove imap4.StringTransportConsumer in favor of StringTransport
and producer management logic.

markrwilliams committed:
imap4.IMAP4Server: test unknown auth method.

Also, use addCleanup to clean up after
AuthenticatorTests.test_customChallengers.

markrwilliams committed:
imap4.IMAP4Server: test auth with missing portal.

markrwilliams committed:
imap4.IMAP4Server: test an auth challenger that raises an exception.

markrwilliams committed:
imap4.IMAP4Server: test an undecodable challenge response.

markrwilliams committed:
imap4.IMAP4Server: test unhandled credentials.

markrwilliams committed:
IMAP4Server failure response message assertion becomes a method.

This de-duplicates redundant assertions.

markrwilliams committed:
Correct failed IMAP4Server response type in test docstrings.

markrwilliams committed:
imap4.IMAP4Server: test unexpected login exception.

markrwilliams committed:
Merge branch 'trunk' into 9161-py3imap4

markrwilliams committed:
Merge branch '9161-py3imap4' of github.com:twisted/twisted into 9161-py3imap4

markrwilliams committed:
imap4.IMAP4Server: test two STARTTLS commands.

markrwilliams committed:
imap4.IMAP4Server: test STARTTLS with existing challengers.

markrwilliams committed:
imap4.IMAP4Server: test LOGIN received before STARTTLS.

markrwilliams committed:
imap4.IMAP4Server: test missing portal and use base class' auth.

twisted.mail.test.test_imap.SimpleServer's theAccount is now always a
class variable, and TestRealm can now retrieve it from SimpleServer on
each requestAvatar invocation.  This allows the removal of
SimpleServer.authenticateLogin, so that tests can exercise
IMAP4Server.authenticateLogin directly.

This also tests what happens upon receipt of LOGIN without a portal.

markrwilliams committed:
imap4.IMAP4Server: test when a checker does not return IAccount.

markrwilliams committed:
imap4.IMAP4Server: test unexpected exception during login.

markrwilliams committed:
imap4.IMAP4Server: test mailboxes without namespaces.

markrwilliams committed:
imap4.IMAP4Server: test selecting a mailbox that does not exist.

glyph commented:
May I suggest prioritizing pyflakes fixes so that the tests start passing (periodically) again?  I don't think the issue is fully fixed but maybe this 4kloc branch should just land at this point, and additional test additions could be smaller and slightly more reviewable?

markrwilliams commented:
@glyph I see two choices: we can land this after satisfying pyflakes, or I can split this into at least two PRs: one that adds tests, and one that actually ports `imap4` to Python 3.

I'd like to split this up for a couple of reasons:

1. No one is going to be able to review a PR this large.
2. We need 100% test coverage before we can say the port actually works.
3. There are unrelated bugs fixed in this branch that should be in their _own_ PRs, like [9207](https://github.com/twisted/twisted/pull/805/commits/8f5ac1a0c2a543ea0d088c52bd068e49380d719a).

Running `twisted.mail.test.test_imap` from this branch against `twisted.mail.imap4` from trunk results in 7 failures and 7 errors, so I don't think it's going to be a terrific amount of effort to split this up.  What d you think?

markrwilliams committed:
imap4.IMAP4Server: uncomment inexplicably commented-out test code.

markrwilliams committed:
test_imap: Remove unused import.

markrwilliams committed:
Merge branch 'trunk' into 9161-py3imap4

glyph commented on src/twisted/mail/imap4.py:
```diff
@@ -4784,26 +5035,35 @@ def collapseNestedLists(items):
     for i in items:
+        if isinstance(i, unicode):
+            # anything besides ASCII will have to wait for an RFC 5738
+            # implementation.
```
Could this be a link to a Twisted ticket?

markrwilliams replied:
https://twistedmatrix.com/trac/ticket/9258 !

glyph commented on src/twisted/mail/test/test_imap.py:
```diff
@@ -5348,6 +7242,108 @@ def timedOut(failure):
+        # The loopback method's Deferred fires the connection is
+        # closed, and the server closes the connection as a result of
+        # expireTime.
+        return defer.gatherResults([d, self.loopback()])
```
It is somewhat disheartening to see so many new tests returning `Deferreds` when they really shouldn't need to; however, I can see that it's following the local convention in this test suite, and it's better to have coverage than not.  Can you file a ticket to replace these with `IOPump` and make them not return `Deferred`s?

markrwilliams replied:
https://twistedmatrix.com/trac/ticket/9259 !

glyph commented on tox.ini:
```diff
@@ -100,7 +100,7 @@ commands =
     ; Run the tests wrapped using coverage.
     {withcov,coverage}: python {toxinidir}/admin/_copy.py {toxinidir}/admin/zz_coverage.pth {envsitepackagesdir}/zz_coverage.pth
-    {withcov,coverage}: coverage erase
+;    {withcov,coverage}: coverage erase
```
Please revert this and let's discuss whether it's appropriate on a separate ticket :).

markrwilliams replied:
Whoops - that should not be there!

glyph commented on src/twisted/python/_setup.py:
```diff
@@ -364,10 +364,15 @@ def _checkCPython(sys=sys, platform=platform):
 _isCPython = _checkCPython()
 
 notPortedModules = [
+    "twisted.internet.glib2reactor",
```
Why are we *adding* to this list in this branch?

glyph replied:
To make this more actionable: if this is not necessary in this branch, please revert this hunk out.

markrwilliams replied:
This is some kind of merge issue :(

markrwilliams committed:
Merge branch 'trunk' into 9161-py3imap4

markrwilliams committed:
Revert inadvertent changes.

markrwilliams committed:
imap4.IMAP4Client: complete incomplete comment.

markrwilliams committed:
imap4: comment on implementation limitation links to ticket.

markrwilliams committed:
Merge branch 'trunk' into 9161-py3imap4

rodrigc commented:
@markrwilliams I can confirm that with this fix, the IMAP example on the front page of https://twistedmatrix.com now works with Python 3.6. (I tested with *strport="ssl:imap.gmail.com:993"*)

Do you have edit permissions on the front page?  In that example, the *user* and *password* need to be prefixed with *b* for bytes.  Other than that, the example works.  It would be good to fix that example on the wiki.

markrwilliams commented:
@rodrigc Thank you so much for your thoroughness!  I edited the front page - please check I got it right!

rodrigc commented:
Looks good.
If I replace **strport="ssl:imap.gmail.com:993"** with **strport="tls:imap.gmail.com:993"**,
I get an error: `IMAP4 client interaction failed [('SSL routines', 'tls_process_server_certificate', 'certificate verify failed')]`

I did a trace with wireshark and found this error:

**TLSv1.2 93 Alert (Level: Fatal, Description: Unknown CA)**

Is that a problem on gmail's side, or in the Twisted code?

markrwilliams commented:
@rodrigc I think `tls` not working is an issue with your OpenSSL installation, but I think `ssl` a bad endpoint and we should get rid of it.  It does not verify server certificates against _any_ trust root unless you specify a directory of CA certificates (e.g. `ssl:imap.gmail.com:993:caCertsDir=/etc/ssl/certs`) or a hostname (e.g., `ssl:imap.gmail.com:993:hostname=imap4.gmail.com`). In the former case, it will read CA certs out of the provided directory, while in the latter case it use [`t.i.ssl.platformTrust`](https://twistedmatrix.com/documents/current/api/twisted.internet.ssl.platformTrust.html) to determine the location of those certs.

It will also not offer SNI or perform SNI validation without the `hostname` keyword argument.

This is a needlessly complicated API with unsafe defaults.  It's so bad that the same [issue](http://bugs.python.org/issue21308) [prompted](https://www.python.org/dev/peps/pep-0466/) the release of [Python 2.7.9](https://www.python.org/downloads/release/python-279/).

Regardless, `tls` verifies that a trusted CA certificate has signed a server's certificate, whereas `ssl` does not.  Upgrading your system's trust store - `security/ca_root_nss`  on FreeBSD, `ca-certificates` on Debian-based distributions - might help, but you might also have to upgrade OpenSSL.

Let me know if that works!

rodrigc commented:
I am using OS X, and `/etc/ssl/certs` does not exist.  Also `/usr/local/etc/openssl/certs`
is empty.  So trying this did not work.  There is some discussion about this in:

https://github.com/twisted/twisted/commit/71282ea1076374e04c5bcc772fc12d989021df27
https://twistedmatrix.com/trac/ticket/6372
https://twistedmatrix.com/trac/ticket/5446

I don't know how to get this to work.

markrwilliams commented:
@rodrigc I don't have a Mac, but I poked around inside a few wheels and confirmed with @alex that `pyopenssl` reads `/usr/local/etc/openssl/certs` via `cryptography`, and that this directory can be populated by installing OpenSSL with brew.

You might need to upgrade `pyopenssl` and `cryptography` as well.

Let me know if all this makes the `tls` endpoint work.

rodrigc commented:
@alex do you know the steps required to get the Twisted tls endpoint to work on Mac?

It would be good to have a web link with the steps needed to get the tls endpoint
to work on Mac (if we don't have one already).  Also, it would be good to refer to this link from the *Mail Client* example web page.

I think that many people use Mac to code things in Python,
and it is quite reasonable to assume that many newcomers to Twisted who use
Mac will copy and paste the *Mail Client* example to try things out.

glyph commented:
The caveats for macOS are documented here: https://twistedmatrix.com/trac/ticket/9209
